docs/pallas/tpu/distributed.md

```diff
@@ -0,0 +1,1453 @@
+- `device_id_type` specifies the format of `device_id`, which can either be in LOGICAL format (integer device ID), or in MESH format (an ND-tuple index into the logical device mesh). The default mode is MESH.
+
+`make_async_remote_copy` returns a descriptor object on which you use the `.start()` method to initiate the DMA, and the `.wait_send()` to block on `send_sem` and `.wait_recv()` to block on `recv_sem` (or `.wait()` to block on both). If a device is only expected to send data, it is sufficient to only call `.start()` and `.wait_send()`, and likewise if a device is only receiving it is sufficient to only call `.wait_recv()`. If using a SIMD pattern where all devices execute the DMA, each device will generally call both `.start()` and `.wait()`.
+```python
+dma_descriptor = make_async_remote_copy(src_ref, dst_ref, send_sem, recv_sem, device_id)
+dma_descriptor.start() # Initiate the DMA (non-blocking).
+# ... do other work
+dma_descriptor.wait_send() # Block until all data has been sent.
+dma_descriptor.wait_recv() # Block until all data has been received.
+```
+
+As an example, let's visualize a DMA where we consider 4 devices (indexed 0, 1, 2, 3). We consider a scheme where device 0 copies to device 1, and device 2 & 3 copy to each other. In practice, we can create such an asymmetric communication pattern by using `@pl.when` to branch on the device ID.
+
+(1) Each device creates the DMA descriptor. Devices 0, 2, and 3 call `.start()` to initiate the DMA from `src_ref`. Device 1 is idle.
+
+![rdma_start](../../_static/pallas/distributed/rdma_start.png)
+
+(2) As `.start()` is non-blocking, each device is free to do other computation while the DMA is in flight. Devices 0, 2, and 3 call `.wait_send()` to wait on `send_sem` which blocks until all data has been sent.
+
+![rdma_send](../../_static/pallas/distributed/rdma_send.png)
+
+(3) Finally, devices 1, 2, and 3 will call `.wait_recv()` to wait on `recv_sem` until all data has arrived at `dst_ref`.
+
+![rdma_recv](../../_static/pallas/distributed/rdma_recv.png)
+
+### DMA Semaphores
+
+`send_sem` and `recv_sem` are instances of a special type of semaphore reserved exclusively for use with DMAs. They must be allocated with the `tpu.SemaphoreType.DMA` type when specifying input specs to `pallas_call`.
+
+Internally, DMA semaphores can be thought of as progress trackers. On DMA start, the local device will initialize its own `send_sem` and the receiver's `recv_sem`. The value of the semaphore is asynchronously updated while the DMA executes. Waiting on each semaphore will block until the value of the semaphore reaches the total amount of data sent/received, and will be decremented by that amount when the value is reached. This means that either all data has been sent (for `send_sem`) or all data has been received (for `dst_sem`). The value of the semaphore can be read with `pl.semaphore_read`, but the underlying semantics of the value could change between hardware generations.
```

Comment:
```suggestion
Internally, DMA semaphores can be thought of as integer-valued progress trackers. On DMA start, the local device will initialize its own `send_sem` and the receiver's `recv_sem` to zero. The value of the semaphore is asynchronously incremented while the DMA executes. Waiting on a semaphore will block until the value of the semaphore reaches the total bytes of data sent/received; when the value is reached, waiting threads are released and the sempahore's value is reset back to zero. This means that either all data has been sent (for `send_sem`) or all data has been received (for `dst_sem`). The value of the semaphore can be read with `pl.semaphore_read`, but the underlying semantics of the value could change between hardware generations.
```

Are these edits accurate? Trying to make it more precise.

Reply:
Updated to:
"Internally, DMA semaphores can be thought of as integer-valued progress trackers. On DMA start, the local device will begin to increment the value of `send_sem` and the receiver's `recv_sem` asynchronously. Waiting on a semaphore will block until the value of the semaphore reaches the total bytes of data sent/received; when the value is reached, waiting threads are released and the sempahore's value is decremented by the same amount. This means that either all data has been sent (for `send_sem`) or all data has been received (for `dst_sem`). The value of the semaphore can be read with `pl.semaphore_read`, but note that the underlying semantics of the value could change between hardware generations (e.g. the value may not represent exactly the number of bytes sent, although this is a useful mental model to have when reasoning about the behavior of the semaphore).."

Semaphores in general aren't reset to zero but are decremented by the number of bytes sent - if there is a bug the semaphore could be left in a non-zero state when the program exits. Also the actual value isn't necessarily the number of bytes sent (this is not defined in a spec anywhere) so added a note about that.